_posts/2019-12-13-green-sock-animation.md

```diff
@@ -0,0 +1,114 @@
+<div style="background-color: #000000; width: 50px; height: 50px" id="log_fade"></div>
+</div>
+
+While the fade out may look pretty similar, when fading in, the square on the left seems to appear almost instantly, and slowly gain its last bit of opacity, while the square on the right seems to be an even fade. In reality, the left square is a linear transition, while the right square is exponential. i.e. the graphs of their opacity over time look something like:
+
+<img src="/assets/images/articles/opacity_graph.png" alt="A graph of opacity over time">
+
+Achieving this kind of control over keyframes _probably_ exists, but I couldn't find it in a cursory google search, so I moved on to bigger guns for my solution.
+
+## Enter GreenSock
+
+So what is [GreenSock](https://greensock.com/) exactly? To quote them directly:
+> Ultra high-performance, professional-grade animation for the modern web
+
+I personally believe the "high-performance" is both an attribute of the speed of the library itself, and also the speed of development with the library. I spent so much time trying to figure out how to get css to animate things the way I wanted to, and it took me probably a quarter of the time to get a GreenSock version of the same thing (only better) running.
+
+One of the greatest assets of GreenSock, in my opinion, is the timeline feature. This makes coordinating multiple animations a breeze. If we consider the previous example, using keyframes, doing the same thing with GreenSock is _way_ simpler:
+
+```html
+<style>
+#bg-image, .buttons {
+    opacity: 0
+}
+</style>
+<script>
+$(function () {
+let tl = gsap.timeline()
+tl.to("#bg-image", {opacity: 1, duration: 3})
+tl.to(".buttons", {opacity: 1, duration: 3}, "-=1")
+})
+</script>
+```
+That was all it took to get the same functionality as the keyframes. We create a GreenSock timeline (`tl` here) and we then start adding actions to it. We give it a css selector, and then our arguments, which are essentially our destination. We set our css to opacity 0 so that at page load, the elements are hidden, and then GreenSock's `to` functionality handles the tweening to get the element to its destination (opacity of 1). 
+
+Timelines by default are sequential, so the next action doesn't start until the previous one has finished, but we can add modifiers to change this. `"-=1"` is one such modifier. This is telling GreenSock to start this animation 1 second _before_ the previous action finishes, so they overlap. So now how do we fix the fact that the fade doesn't look linear? Simple, we add an ease, which GreenSock already has built in. To add an ease, all we have to do is add it as another argument:
+
+```javascript
+tl.to("#bg-image", {opacity: 1, duration: 3, ease: 'power2.in'})
+tl.to(".buttons", {opacity: 1, duration: 3, ease: 'power2.in'}, "-=1")
+``` 
+And that is it. We have overlapped our fades, and they look like they fade in evenly. Easy.
+
```

Comment:
Could add a disclaimer on licensing with Greensock maybe? If you're making a product to sell to multiple people, not a one off, I believe you need licensing:
https://greensock.com/licensing/

Reply:
🤷‍♂ Kinda feels like a strange addition. I figure, let the greensock website handle telling people that...